### ISSUE TITLE
LSML window cannot be dragged to a secondary monitor using mouse

### ISSUE THREAD

**Issue by VioletGiraffe** [EmilyBjoerk/lsml#675]
This problem appeared after I upgraded my main monitor to 4K. I used to have two monitors, both with 100% scaling (i. e. no scaling), and there was no problem. Now my primary monitor has 175% scaling and secondary monitor is at 100%. I suspect scaling might be the issue.

I start dragging the window in the direction of the 2nd monitor, but once the LSML window border touches the primary screen border, the window jumps back to the center of the screen.

**Comment by EmilyBjoerk** [comment]
The only multi-screen setup that I have is running Windows 7 / Linux. According to [here](https://superuser.com/questions/29796/dual-monitors-on-windows-how-do-i-set-a-different-dpi-or-text-size-on-each-mon) it is not possible to set different text scale for each monitor or windows 7 so I'm unable to debug or test this. 

Also what should happen if LSML is half on one screen and half on the other with a different DPI? Without applying different scaling on the sides, one side will look "smaller". And it's impossible for me to apply different scaling on different parts of the window, this is something that JavaFX has to handle internally (and I'm not sure it does).

At the current state I'm unable to fix this issue. I'm sorry.

**Comment by VioletGiraffe** [comment]
> what should happen if LSML is half on one screen and half on the other with a different DPI? Without applying different scaling on the sides, one side will look "smaller"

Yes. One side would either be smaller or larger, but that's not something you should be concerned with. It's handled internally either by Windows or by JavaFX. Based on the blurry fonts, I infer that either JavaFX doesn't support high DPI, or your application doesn't enable this support, so Windows is just up-scaling all bitmaps. Either way, it is passable, my only complaint is not being able to move the windows around freely. It literally jumps back once the border of the window touches the border of the monitor.

**Comment by EmilyBjoerk** [comment]
As far as LSML goes, all dimensions in the UI are specified in device independent units (pt, em, % of parent etc) and automatic layout is used everywhere. All graphical elements are vector graphics so it should all just work on high DPI displays. On my high DPI laptop the fonts and all graphics are rendered laser sharp even with high scaling. So it's definitely working, I think the issue is Java not handling mixed DPI displays under Windows 10.

At any rate I'm unable to test this as I don't have Windows 10 which seems to be needed to set different DPI on each screen. Would you be able to debug the code? I can point to where you need to look.

**Comment by VioletGiraffe** [comment]
>Would you be able to debug the code? I can point to where you need to look.
I can try. Please do provide pointers, and I'll look into it when I have the time to setup Java dev environment on the machine where I can reproduce this (which is unfortunately not my main PC - that only has one monitor).

**Comment by EmilyBjoerk** [comment]
I think that the instructions [on the wiki](https://github.com/EmilyBjoerk/lsml/wiki/How-to-Contribute) are mostly up to date. Try them out and if it doesn't work, file a new issue against the wiki page and I will take a look.

**Comment by EmilyBjoerk** [comment]
@VioletGiraffe did you ever get around to debugging this?

**Comment by VioletGiraffe** [comment]
I didn't, but if you're back and are going to update LSML, I'll be happy to have a look.

**Comment by EmilyBjoerk** [comment]
Please do :)

**Comment by VioletGiraffe** [comment]
The mouse coordinates look alright, there's most likely no problem in the `onMouseDragged` method. It's as if the application doesn't know there is a second monitor available, and simply resets the window to the center of the screen as I attempt to cross the monitor border onto the adjacent monitor.

**Comment by EmilyBjoerk** [comment]
Since I now have W10, I did some testing by forcing my primary display to 175% scaling and 100% on the secondary. 

I am able to move the window across the displays and the scaling is applied correctly, JavaFX even renders the SVG images sharply on the scaled display (without up scaling the bitmaps). Tried with the scaled monitor to both the left and right. Both worked, however launching the jar directly had a different behaviour than running the exe.

However there is a dead-zone when moving between the displays that is proportional to the scaling factor difference. Also what I do see is that the grabbing point changes on the LSML window once the scaling is applied. So by keeping the mouse button down while moving back and forth across the scaling border the cursor position will gradually shift to the right. 

Maybe your geometry is such that this shifting is causing issues or the deadzone is just bad. Also I understand that W10 treats the app differently depending on if manifest of the exe sets a "High DPI awareness flag" or not. The lsml.exe file to my knowledge doesn't set this. But launching the jar directly might set it (depending on the manifest of your JVM). 

I'm not sure what I can do on LSML's side. I guess reworking the resizing and window moving logic to not use delta changes in some way might get rid of the cursor shift but I can't say if this would fix the issues you are having. 

Unfortunately this is a case of "works for me" and I don't have the hardware to do a better reproduction of the issue :/

**Comment by VioletGiraffe** [comment]
Do you want me to record the issue on video? Perhaps, then you'll better understand what's going on.
Also, I did not notice a difference between .exe and launching from Eclipse directly .

**Comment by EmilyBjoerk** [comment]
I think a video would be useful. But based on what I saw on my testing, I'm not sure if I can do anything through code...

Are other javafx apps also having issues?

**Comment by VioletGiraffe** [comment]
I don't know about any other JavaFX application, can you point me to one? I could always run some sample, of course, but it might not be the same as a full-fledged product.

**Comment by EmilyBjoerk** [comment]
You could try some sample apps from the javafx docs (sorry, I don't have a link atm).

Javafx doesn't expose any information about display scaling (probably because it doesn't make sense on systems that use display DPI to properly scale stuff). AFAIK (I haven't dug in really deep) I only see a flat coordinate space that covers both displays and the rest is handled by the jdk. I suspect that for applications that use the normal window decorations it'll work just fine but LSML uses a border less window class and thus has to handle window manipulation gestures manually. Given the information exposed to me by the javafx API I don't see what I can do other than changing from relative positioning to absolute positioning but like I said I'm not sure that'll help.

**Comment by LiSongMWO** [comment]
I now have a 4K + 2K display and I can repro this... 3 years later xD

**Comment by VioletGiraffe** [comment]
Better late than never! Would be great to have LSML updated and be worked on once again :)

**Comment by LiSongMWO** [comment]
Uh... well I *HAD* a repro... I updated my JDK and now I can't repro anymore no matter what I try... I think this is probably a bug in the JavaFX library and was probably fixed in one of the later JDK versions... 

I'm going to assume it's a bug in Java based on this, please try the latest version of Java 8 (8u301 or later). I'll close for now, if it's still an issue, please re-open.

**Comment by VioletGiraffe** [comment]
Would it be difficult for you to provide a binary build? Or maybe set up Appveyor or Github build job with a build artifact that we can download?

**Comment by LiSongMWO** [comment]
I'm only maintaining right now. I have already spend way more time than I wanted to while fixing the parsing and RAC bugs for this release so I don't want to set up anything extra.

If you want to run the latest code, just check out the git repo, make sure you have Java **8** as your JRE/JDK on your PATH. And run `gradlew.bat run` to run the code. 

2.0.12 will be out in a few hours assuming I can get the build system to work.

**Comment by VioletGiraffe** [comment]
Gotcha, thanks. It's great to know Eclipse is no longer required for building (or maybe it never was and I just didn't realize).

**Comment by LiSongMWO** [comment]
Eclipse was never required for running, you could always run with the bundled gradle wrapper it was the preferred IDE but I'm using intellij now so 🤷
